### DISCUSSION TITLE
SEP-0007: Variable Substitution

### DISCUSSION THREAD

**Discussion by afs** [w3c/sparql-dev#169] [Ideas]
Discussion for _[SEP-0007](https://github.com/w3c/sparql-12/blob/main/SEP/SEP-0007/sep-0007.md): Variable Substitution_.

**Comment by afs** [comment]
"substitution" is the terminology used in the SPARQL spec. But direct replacement has problems. The solution proposed to is put in inline data and join it with the pattern, where patterns can bind variables (Basic Graph Patterns, Property Path Expressions, Graph)

For `(?x = 23)`, the BGP `{ ?s :p ?x }` becomes `{ ?s :p ?x . VALUES ?x { 23 } }`.

Would `inject` be a better name? There is no substitution/replacement going on.

(There are several different ways to put in the values, and the order of the join can be reversed because the inline data only occurs when a variable will cause a binding.)

**Comment by namedgraph** [comment]
Is `ParameterizedSparqlString` using direct replacement? Will it be affected if these changes are implemented?

**Comment by afs** [comment]
Parametrization is different.

SEP-0007 is for LATERAL joins and EXISTS. 

Please ask on the Apache Jena users mailing list (or read the source code) about impact.
